### ISSUE TITLE
Review and add redundancy to data feed

### ISSUE THREAD

**Issue by LindsayYoung** [18F/openFEC#1716]
It seems like there are some transactions that are slipping though the cracks on the large, schedule tables.

The current issues will be solved with partitioning since we are rebuilding those tables but the data integrity is extremely important so: 

    1) we should identify the source of the missing data
    2) we should add some extra processes to check and fix missing data

We might want to do 2 first to help us identify when the problems are occurring and make sure the data is always accounted for.

**Comment by noahmanger** [comment]
Another instance of missing data: https://github.com/18F/openFEC/issues/1720

**Comment by LindsayYoung** [comment]
I think that person downloaded filings and then did not find filings again. This is data in a FEC table that is not in a ofec_ table.

**Comment by noahmanger** [comment]
It's unclear what they did or didn't try to do. I think they're point about downloading receipts was potentially mistaken. All I know is that that committee has reported itemized receipts but they're not showing up.

**Comment by LindsayYoung** [comment]
OK, thank for clarifying I just read their part before

**Comment by PaulClark2** [comment]
I agree with Noah's comment. However, I can confirm there is a difference however between beta and the presidential map for contributions from New Mexico to Mr. Trump. The map and the DB (PG sched_a and FECP & DW sched_a) have 53 Trump contributions from New Mexico `select * from sched_a s where s.cmte_id = 'C00580100' and s.contbr_st = 'NM';`  Beta only has 27 contributions from New Mexico.

**Comment by noahmanger** [comment]
Criteria for completion: 
- [ ] We have a backup system for transfer of records
- [ ] We have a system for alerting us when our counts get out of line

**Comment by noahmanger** [comment]
So is the only step on this to add the alerting system? (Re criteria for completion above).

**Comment by LindsayYoung** [comment]
Correct. We still want to use counts to flag any discrepancies going forward. I have not started that yet

**Comment by ccostino** [comment]
Given that we just switched tables off of the data warehouse versions, these alerts should be based on count discrepancies between the `fec-vsum_sched_*` tables and our `ofec_sched_*` versions, correct?  Or is there something else we should be basing the counts off of instead?

Also, should these alerts take place along with our refresh task, be an entirely separate check, or hang off of another event like the nightly updates?

**Comment by LindsayYoung** [comment]
I was thinking that the check could be done weekly before we rebuild the assets, I think it would be sufficient to do the schedule records that were uploaded for the week + new records - records to be deleted and see if that is the same count you get in the corresponding ofec table after `refresh_and_rebuild` is run. Large counts can be a little hard on the RDS boxes, so we might just want to count what was imported for the week, though the dates have indexes in the fec_vsum tables so maybe that would not be more efficient.

**Comment by ccostino** [comment]
Ah, perfect!  Yes, let's do the report before the asset rebuild each week.  The counts are definitely a bit troublesome given how much time it takes, even with the resources we have allocated on the RDS boxes now.  As for keeping track of the schedule records that were uploaded for the week, we don't have a means of knowing what those are currently, right?  Those updates occur (or are supposed to occur!) nightly, so those queue tables are cleared out once the updates are done.  We'd need to build some sort of support in place to keep track of how many records were supposed to be added/removed for the week.

Assuming we had that, the equation/check would look like this if I read everything correctly:  `(number of records added - number of records deleted for the week) + (number of records to be added - number of records to be deleted tonight)` == number of records in corresponding ofec table.`

Unfortunately, the `fec_vsum_sched_a` and `fec_vsum_sched_b` tables do not have indexes on any of the date fields, just the primary key field (`sub_id`).  However, the primary key is truly a unique identifier in these tables now so that is at least reliable for us in terms of record count, but if we can't make use of the index on the table the query takes about 10 minutes to run (I just ran a test in dev).  I'm trying to figure out if it's possible to get it to make use of the index (https://wiki.postgresql.org/wiki/Index-only_scans).

**Comment by LindsayYoung** [comment]
I think we are saying the same thing, but I am slightly confused because I don't think we need to keep track of the que over the week. We can just check once a week and see if the counts are off. Tell me if I am missing something.

I was thinking, once a week we can:
1) count what is in the ofec table, taking into account what has not been processed yet with:
ofec_total = count records in ofec_sched_a with pg_date in the last week + things waiting in the new que - things waiting in the old que
2) run the aggregate rebuild 
3) new_ofec_total =  count the ofec_total with pg_date in the last week
4) compare the count and email 

A 10 minute process is less than ideal, but worth it if we find data update errors

**Comment by ccostino** [comment]
Okay, now I see!  I'm sorry for the confusion, I had overcomplicated it by not taking into account the `pg_date` column and thinking we'd have to cross reference against the source table.

If we're just focusing on the amount of records within the last week, this should be easier and quicker.  We could add an index on the `pg_date` column and that should help improve performance. :-)

**Comment by noahmanger** [comment]
Can this be closed now?

**Comment by ccostino** [comment]
I believe so!
